### ISSUE TITLE
wrong calculation of visibility by domain and product visibility date

### ISSUE THREAD

**Issue by milanStano** [shopsys/shopsys#1924]
### What is happening

If I set the end of the sale date to today’s in the administration and save it, then the visibility of the product is recalculated on individual domains. But the calculation uses new Datetime() and the date from the form, since I only set the date, it returns with the time 00:00:00, and since we have a timezone set on the front end to Prague, this date is changed to UTC, ie. - 2 hours. So this date will become yesterday for me and so the product will be hidden, which is not right and should remain visible. This can also be a problem with other dates where the time is not set.

Why not set timezone for domain in domain.yml ? 
### Expected result

Calculate visibility must be by domain time zone. (https://github.com/shopsys/shopsys/blob/master/packages/framework/src/Model/Product/ProductVisibilityRepository.php#L183)
 ```
/**
     * @return \DateTime
     */
    public function createDateWithTimezoneCorrection(): \DateTime
    {
        $dateByTimeZoneProvider = new \DateTime('midnight', $this->getDisplayTimeZone());
        $dateByTimeZoneProvider->setTimezone(new \DateTimeZone('UTC'));

        return $dateByTimeZoneProvider;
    }



```

**Comment by pk16011990** [comment]
Working with datetime is confusing for more developers. 

I have few suggestions:
- deny using of `date` column type in Doctrine
- create helpers for common use-cases (filtering by date needs to datetimes instead of "now")
- fix all usages in framework/project-base
- documentation
  - pitfalls
  - cron modules are delayed by timezone (critical for modules like donwloading ratios ČNB)
  - why do not use `date` column type in Doctrine
  - how filter by date in DB
  - how use visibility by date in different timezone than UTC (New Zeland doesn't want to (un)hide products etc. in 12PM)

**Comment by grossmannmartin** [comment]
for denying using certain functions and methods can be used https://github.com/spaze/phpstan-disallowed-calls

**Comment by s3tezsky** [comment]
fixed in #2006
